### PR TITLE
WIP enable northd with etcd-ovsdb

### PR DESCRIPTION
**Pull request by msherif1234**
Signed-off-by: Mohamed Mahmoud <mmahmoud@redhat.com>

### CODE REVIEW

**Comment by numansiddique** [comment]
Thanks @msherif1234 for this PR.   I think it will be great to enhance the "build" command to include ovsdb-etcd in the image.

File: ovn_cluster.sh [hunk 1]
```diff
@@ -446,6 +446,39 @@ function start-db-cluster() {
     $SSL_ARGS start_northd
 }
 
+function check_etcd_mode() {
+    if [[ ! -z "$use_etc" ]]; then
+	    true
```
Comment (flavio-fernandes) [inline ovn_cluster.sh:451]: NIT: space is inconsistent

File: ovn_cluster.sh [hunk 2]
```diff
@@ -446,6 +446,39 @@ function start-db-cluster() {
     $SSL_ARGS start_northd
 }
 
+function check_etcd_mode() {
```
Comment (flavio-fernandes) [inline ovn_cluster.sh:449]: NIT: How about we rename this function to nb_etcd or something other than "check"?

File: ovn_cluster.sh [hunk 3]
```diff
@@ -446,6 +446,39 @@ function start-db-cluster() {
...
+    if [[ ! -z "$use_etc" ]]; then
+	    true
+    else
+        false
+    fi
+}
+
+function start_etcd() {
+    name=`hostname -s`
```
Comment (flavio-fernandes) [inline ovn_cluster.sh:458]: NIT: please use $() instead of back-ticks:   name=$(hostname -s)

File: ovn_cluster.sh [hunk 4]
```diff
@@ -446,6 +446,39 @@ function start-db-cluster() {
...
+
+    rm -rf ${name}.etcd
+
+    ip=`hostname -i | awk '{print $2}'`
```
Comment (flavio-fernandes) [inline ovn_cluster.sh:462]: ip=$(hostname -i | awk '{print $2}')

File: ovn_cluster.sh [hunk 5]
```diff
@@ -446,6 +446,39 @@ function start-db-cluster() {
...
 
+function check_etcd_mode() {
+    if [[ ! -z "$use_etc" ]]; then
+	    true
+    else
+        false
+    fi
+}
+
+function start_etcd() {
```
Comment (flavio-fernandes) [inline ovn_cluster.sh:457]: Q: We are going to need a special version of ovsdb to talk with etcd in one side and northd on the other, right? Is that done in a separate function?

File: ovn_cluster.sh [hunk 6]
```diff
@@ -514,7 +547,12 @@ function start() {
             start-db-cluster
             central=${CENTRAL_NAME}-1
         else
-            ${RUNC_CMD} exec ${CENTRAL_NAME} ${OVNCTL_PATH} start_northd
+            if check_etcd_mode; then
+                start_etcd
+                ${RUNC_CMD} exec ${CENTRAL_NAME} ${OVNCTL_PATH} --ovn-manage-ovsdb=no start_northd
```
Comment (flavio-fernandes) [inline ovn_cluster.sh:552]: Special attention will also be needed when stopping northd, right?

File: ovn_cluster.sh [hunk 7]
```diff
@@ -446,6 +446,39 @@ function start-db-cluster() {
...
+    ip=`hostname -i | awk '{print $2}'`
+
+    etcd --name ${name}  --initial-advertise-peer-urls http://${ip}:2380 \
+     --max-txn-ops 1000000 \
+     --max-request-bytes 17179869184 \
+     --listen-peer-urls http://${ip}:2380 \
+     --listen-client-urls http://${ip}:2379,http://127.0.0.1:2379 \
+     --advertise-client-urls http://${ip}:2379 \
+     --initial-cluster-token etcd-cluster-1 \
+     --initial-cluster ovn1=http://10.126.111.40:2380,ovn2=http://10.126.111.52:2380,ovn3=http://10.126.111.66:2380 \
```
Comment (flavio-fernandes) [inline ovn_cluster.sh:471]: Please make sure to look at ovn_cluster.sh fo use variables instead of hard coded ips.

**Review by flavio-fernandes — COMMENTED** [review]
Very interesting! I'm curious to try this as you get it going.

Is this going to be using  https://github.com/IBM/ovsdb-etcd ?

**Comment by msherif1234** [comment]
> Very interesting! I'm curious to try this as you get it going.
> 
> Is this going to be using https://github.com/IBM/ovsdb-etcd ?

correct

File: ovn_cluster.sh (hunk 6 above)
Comment (msherif1234) [reply ovn_cluster.sh:552]: > Special attention will also be needed when stopping northd, right?

correct but I am not there yet, just trying to get it working

File: ovn_cluster.sh (hunk 2 above)
Comment (msherif1234) [reply ovn_cluster.sh:449]: is_ovsdb_etcd_mode() any better ?

File: ovn_cluster.sh (hunk 5 above)
Comment (msherif1234) [reply ovn_cluster.sh:457]: yes what we have here is the server side I am adding the client side

File: ovn_cluster.sh (hunk 2 above)
Comment (flavio-fernandes) [reply ovn_cluster.sh:449]: > is_ovsdb_etcd_mode() any better ?

yes!

**Comment by msherif1234** [comment]
will open when it more ready

**Pull request closed without merging**